### PR TITLE
fix: check for models before deleting workspace

### PR DESCRIPTION
**Pull request by kkunapuli**
## Description

Details in [DET-9982]. Previously, if a user tried to delete a workspace with models (archived or not), they were met with an SQL error. Most of the workspace contents were deleted and the workspace was left in state "Deleting". 

With this change, we first check if a workspace has models on a delete request. If it does, a human-readable error message is returned and the workspace state is unchanged. Nothing in the workspace is deleted.

## Test Plan

Added integration tests to confirm functionality. Also tested locally using CLI. 

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
DET-9982

[DET-9982]: https://hpe-aiatscale.atlassian.net/browse/DET-9982?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/api_workspace.go [hunk 1]
```diff
@@ -551,11 +559,16 @@ func (a *apiServer) DeleteWorkspace(
...
+	}
+
 	holder := &workspacev1.Workspace{}
-	err = a.m.db.QueryProto("deletable_workspace", holder, req.Id)
+	// TODO(kristine): update workspace state in transaction with template delete
+	a.m.db.QueryProto("deletable_workspace", holder, req.Id)
 	if holder.Id == 0 {
-		return nil, errors.Wrapf(err, "workspace (%d) does not exist or not deletable by this user",
-			req.Id)
+		return nil, fmt.Errorf("workspace (%d) does not exist or not deletable by this user", req.Id)
```
Comment (kkunapuli) [inline master/internal/api_workspace.go:571]: Updating error formatting to use `fmt.Errorf` instead of `errors.Wrapf` in this function for two reasons. Mainly because `errors` is deprecated. Also, because the function was relying on `errors.Wrapf` returning `nil` if `err` was `nil`. This way, it is more explicit when the function returns a response and when it does not.

File: master/pkg/model/model.go [hunk 2]
```diff
@@ -4,36 +4,39 @@ import (
 	"fmt"
 	"strings"
 	"time"
+
+	"github.com/google/uuid"
 )
 
 // Model represents a row from the `models` table.
 type Model struct {
-	ID              int       `db:"id" json:"id"`
```
Comment (kkunapuli) [inline master/pkg/model/model.go:11]: I reordered fields to match the postgres output. I can keep the ordering consistent so that changelog is easier to follow if that's preferrable.

File: master/internal/api_workspace.go [hunk 3]
```diff
@@ -119,6 +119,13 @@ func (a *apiServer) getWorkspaceAndCheckCanDoActions(
 	return w, *curUser, nil
 }
 
+func (a *apiServer) workspaceHasModels(ctx context.Context, workspaceID int32) (bool, error) {
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:122]: This change from a users perspective feels a little strange to me.

So deleting a workspace will automatically delete experiments, templates, and NTSC, but not models? Maybe this is something we need product input on, but what do you think about being more consistent and just deleting models attached to the workspace?

File: master/internal/api_workspace.go [hunk 4]
```diff
@@ -551,11 +558,16 @@ func (a *apiServer) DeleteWorkspace(
 		return nil, err
 	}
 
+	modelsExist, _ := a.workspaceHasModels(ctx, req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:561]: we should almost never ignore errors, from the users perspective if there is a database issue on this query returning a different error message can be a really confusing experience for the user

File: master/internal/api_workspace.go [hunk 5]
```diff
@@ -551,11 +558,16 @@ func (a *apiServer) DeleteWorkspace(
...
+	if modelsExist {
+		return nil, fmt.Errorf("workspace (%d) contains models; move or delete models first", req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:563]: Just returning an error like this will make this a 500 http error when it goes through the grpc-go gateway

If you instead return 
```
return status.Errorf(codes.FailedPrecondition, ...)
```

grpc gateway will make this a 400 error

File: master/internal/api_workspace.go [hunk 6]
```diff
@@ -551,11 +558,16 @@ func (a *apiServer) DeleteWorkspace(
...
+	}
+
 	holder := &workspacev1.Workspace{}
-	err = a.m.db.QueryProto("deletable_workspace", holder, req.Id)
+	// TODO(kristine): update workspace state in transaction with template delete
+	_ = a.m.db.QueryProto("deletable_workspace", holder, req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:568]: same comment as above about ignoring errors

File: master/internal/api_workspace.go [hunk 7]
```diff
@@ -551,11 +558,16 @@ func (a *apiServer) DeleteWorkspace(
...
 	if holder.Id == 0 {
-		return nil, errors.Wrapf(err, "workspace (%d) does not exist or not deletable by this user",
-			req.Id)
+		return nil, fmt.Errorf("workspace (%d) does not exist or not deletable by this user", req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:570]: this line is not the same as above we need to specify the error we are wrapping

so for example
```
		return nil, fmt.Errorf("workspace (%d) does not exist or not deletable by this user: %w", req.Id, err)
```

File: master/internal/api_workspace.go [hunk 8]
```diff
@@ -579,19 +591,23 @@ func (a *apiServer) DeleteWorkspace(
 	log.Debugf("deleting workspace %d templates", req.Id)
 	err = templates.DeleteWorkspaceTemplates(ctx, int(req.Id))
 	if err != nil {
-		return nil, errors.Wrapf(err, "error deleting workspace (%d) templates", req.Id)
+		return nil, fmt.Errorf("error deleting workspace (%d) templates", req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:594]: same comment as above about the ```fmt.Errorf``` not being the same as the ```errors.Wrapf```

File: master/internal/api_workspace.go [hunk 9]
```diff
@@ -579,19 +591,23 @@ func (a *apiServer) DeleteWorkspace(
...
 	}
 
 	if len(projects) == 0 {
 		err = a.m.db.QueryProto("delete_workspace", holder, req.Id)
-		return &apiv1.DeleteWorkspaceResponse{Completed: (err == nil)},
-			errors.Wrapf(err, "error deleting workspace (%d)", req.Id)
+		if err != nil {
+			return nil, fmt.Errorf("error deleting workspace (%d)", req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:600]: same comment as above about the ```fmt.Errorf``` not being the same as the ```errors.Wrapf```

File: master/internal/api_workspace.go [hunk 10]
```diff
@@ -579,19 +591,23 @@ func (a *apiServer) DeleteWorkspace(
...
+		}
+		return &apiv1.DeleteWorkspaceResponse{Completed: true}, nil
 	}
 	go func() {
 		a.deleteWorkspace(ctx, req.Id, projects)
 	}()
-	return &apiv1.DeleteWorkspaceResponse{Completed: false},
-		errors.Wrapf(err, "error deleting workspace (%d)", req.Id)
+	if err != nil {
+		return nil, fmt.Errorf("error deleting workspace (%d)", req.Id)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:608]: same comment as above about the ```fmt.Errorf``` not being the same as the ```errors.Wrapf```

File: master/internal/api_workspace.go [hunk 11]
```diff
@@ -579,19 +591,23 @@ func (a *apiServer) DeleteWorkspace(
...
+	}
+	return &apiv1.DeleteWorkspaceResponse{Completed: true}, nil
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:610]:
```suggestion
	return &apiv1.DeleteWorkspaceResponse{Completed: false}, nil
```

File: master/internal/api_workspace_intg_test.go [hunk 12]
```diff
@@ -521,3 +522,29 @@ func TestAuthzWorkspaceGetThenActionRoutes(t *testing.T) {
 		require.Equal(t, expectedErr.Error(), curCase.IDToReqCall(id).Error())
 	}
 }
+
+func TestWorkspaceHasModels(t *testing.T) {
+	// set up a dB and api server to use for integration testing
+	require.NoError(t, etc.SetRootPath("../static/srv"))
+	pgDB, cleanup := db.MustResolveNewPostgresDatabase(t)
+	defer cleanup()
+	db.MustMigrateTestPostgres(t, pgDB, "file://../static/migrations")
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace_intg_test.go:531]:
```suggestion
```

These lines setup a fresh postgres database. It is more expensive from a test execution time perspective and also makes it harder to debug so we shouldn't do this when we don't need to. It would be great if we avoid doing this unless there is a good reason.

GetWorkspaces does this because it is hard to test listing workspaces when other tests can add or there can be preexisting workspaces.

File: master/internal/api_workspace_intg_test.go [hunk 13]
```diff
@@ -521,3 +522,29 @@ func TestAuthzWorkspaceGetThenActionRoutes(t *testing.T) {
...
+
+	// create workspace for test
+	workspaceName := "test-workspace"
+	resp, err := api.PostWorkspace(ctx, &apiv1.PostWorkspaceRequest{Name: workspaceName})
+	require.NoError(t, err)
+
+	// confirm workspace does not have any models
+	exists, err := api.workspaceHasModels(ctx, resp.Workspace.Id)
+	assert.False(t, exists)
+	require.NoError(t, err)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace_intg_test.go:542]: You should always try to check an error before looking at other values a function returns. So I would switch the order of these statements.

File: master/internal/api_workspace_intg_test.go [hunk 14]
```diff
@@ -521,3 +522,29 @@ func TestAuthzWorkspaceGetThenActionRoutes(t *testing.T) {
...
+	api, _, ctx := setupAPITest(t, pgDB)
+
+	// create workspace for test
+	workspaceName := "test-workspace"
+	resp, err := api.PostWorkspace(ctx, &apiv1.PostWorkspaceRequest{Name: workspaceName})
+	require.NoError(t, err)
+
+	// confirm workspace does not have any models
+	exists, err := api.workspaceHasModels(ctx, resp.Workspace.Id)
+	assert.False(t, exists)
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace_intg_test.go:541]: ```assert.False``` will return a boolean if the assertion is true or false? So this line could never fail the test?

I think we want to replace all ```assert.``` with ```require.```?

File: master/internal/api_workspace_intg_test.go [hunk 15]
```diff
@@ -521,3 +522,29 @@ func TestAuthzWorkspaceGetThenActionRoutes(t *testing.T) {
...
+func TestWorkspaceHasModels(t *testing.T) {
+	// set up a dB and api server to use for integration testing
+	require.NoError(t, etc.SetRootPath("../static/srv"))
+	pgDB, cleanup := db.MustResolveNewPostgresDatabase(t)
+	defer cleanup()
+	db.MustMigrateTestPostgres(t, pgDB, "file://../static/migrations")
+	api, _, ctx := setupAPITest(t, pgDB)
+
+	// create workspace for test
+	workspaceName := "test-workspace"
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace_intg_test.go:535]: make this name a ```uuid``` so that this test can run multiple times

File: master/pkg/model/model.go [hunk 16]
```diff
@@ -4,36 +4,39 @@ import (
 	"fmt"
 	"strings"
 	"time"
+
+	"github.com/google/uuid"
 )
 
 // Model represents a row from the `models` table.
 type Model struct {
```
Comment (NicholasBlaskey) [inline master/pkg/model/model.go:12]: I see that this is unused except for tests. Can we just move these models to a test file (maybe like api_checkpoint_intg_test.go) rather than like changing this code even though it is never used?

File: master/internal/api_workspace.go (hunk 3 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:122]: I also wondered about that; I assumed it was because models were considered "high value". I'll check with product on it.

File: master/pkg/model/model.go (hunk 16 above)
Comment (kkunapuli) [reply master/pkg/model/model.go:12]: I don't follow. Are you suggesting we maintain an outdated copy of Model struct for tests, and an updated one for queries?

File: master/internal/api_workspace_intg_test.go (hunk 14 above)
Comment (kkunapuli) [reply master/internal/api_workspace_intg_test.go:541]: `assert.False` asserts that the specified value is false. It passes if `exists` is false and does not if `exists` is true. 

Although, I notice that other tests use require instead of assert; I'll change this to be consistent. I found a nice explanation for the different uses of `assert` and `require`:
``` Using assert will allow the developer to learn about failures on different tests as they occur but it will not halt the rest of the tests execution. But using require will stop the tests on the first failure, providing a fail-fast approach.```

File: master/internal/api_workspace.go (hunk 3 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:122]: Heard back from product here: https://hpe-aiatscale.slack.com/archives/C06GMG83ZE0/p1707234970325459?thread_ts=1707234524.249439&cid=C06GMG83ZE0

**TL;DR** for now, do not delete models along with a workspace as models are one of the most valuable artifacts generated by Determined/MLDE

File: master/pkg/model/model.go (hunk 16 above)
Comment (NicholasBlaskey) [reply master/pkg/model/model.go:12]: I guess my suggestion is that given ```model.Model``` was unused before this change except for tests which really don't use the model efficiently, I don't think this model is worth having and we should delete it or move it to be test only. 

I'd much rather delete the model now so we don't risk it being out of date or becoming out of date. Someone could still add it back when a change will actually use it beyond getting a table name.

File: master/internal/api_workspace.go (hunk 7 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:570]: I did something a little different; I used a separate check for `err != nil` and `holder.Id == 0`. I think it will allow for more sensible error messages; **what do you think?**

File: master/internal/api_workspace.go (hunk 4 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:561]: Makes sense; I misunderstood to what degree we wanted to expose SQL errors to the user. Thanks!

File: master/internal/api_workspace.go (hunk 11 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:616]: I don't understand; isn't the request completed? Earlier, if err == nil we return `Completed: true` right?
from original line 587:
```
if len(projects) == 0 {
		err = a.m.db.QueryProto("delete_workspace", holder, req.Id)
		return &apiv1.DeleteWorkspaceResponse{Completed: (err == nil)},
			errors.Wrapf(err, "error deleting workspace (%d)", req.Id)
```

File: master/pkg/model/model.go (hunk 16 above)
Comment (kkunapuli) [reply master/pkg/model/model.go:12]: The query in `workspaceHasModels` requires an accurate `model.Model`:
```	
exists, err := db.Bun().NewSelect().Model((*model.Model)(nil)).
	Where("workspace_id=?", workspaceID).
	Exists(ctx)
```

I can delete `model.ModelVersion` though if you prefer.

File: master/internal/api_workspace.go (hunk 11 above)
Comment (NicholasBlaskey) [reply master/internal/api_workspace.go:620]: No at this point the request is not completed, notice how we call ```deleteWorkspaces``` in a background goroutine when len(projects) > 0.

```
go func() {
    a.deleteWorkspace(ctx, req.Id, projects)
}()
```

Original line 587 maps to new 602. This is for original line 593.

File: master/pkg/model/model.go (hunk 16 above)
Comment (NicholasBlaskey) [reply master/pkg/model/model.go:12]: Exists queries don't require a model, you can instead give it a table name

```
exists, err := db.Bun().NewSelect().Table("models").
    Where("workspace_id = ?", workspaceID).
    Exists(ctx)
```

I'm also fine if you want to leave the model.

**Comment by NicholasBlaskey** [comment]
Whenever you are ready for me to rereview the whole PR, just request a review from me on github

File: master/internal/api_workspace.go (hunk 7 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:580]: Oh - `holder.Id==0` was used as a substitute for checking for an error?

File: master/internal/api_workspace.go (hunk 11 above)
Comment (kkunapuli) [reply master/internal/api_workspace.go:620]: I understand now; thank you!

File: master/pkg/model/model.go (hunk 16 above)
Comment (kkunapuli) [reply master/pkg/model/model.go:10]: I tried moving it and got stuck in a few import loops. I think there's probably a way to accomplish it but I'm not seeing it and I've spent as much time as I am willing to at this point. 😞

File: master/internal/api_workspace.go [hunk 17]
```diff
@@ -579,19 +598,23 @@ func (a *apiServer) DeleteWorkspace(
...
+		return &apiv1.DeleteWorkspaceResponse{Completed: true}, nil
 	}
 	go func() {
 		a.deleteWorkspace(ctx, req.Id, projects)
 	}()
-	return &apiv1.DeleteWorkspaceResponse{Completed: false},
-		errors.Wrapf(err, "error deleting workspace (%d)", req.Id)
+	if err != nil {
+		return nil, fmt.Errorf("error deleting workspace (%d): %w", req.Id, err)
+	}
```
Comment (NicholasBlaskey) [inline master/internal/api_workspace.go:616]:
```suggestion
```

err will always be nil at this point, this is checking the error from 599

File: master/pkg/model/model.go (hunk 16 above)
Comment (NicholasBlaskey) [reply master/pkg/model/model.go:10]: oh yeah sorry I didn't realize it was used in ```internal/db``` tests

internal imports from db so if you put it in ```internal/db/postgres_test_utils.go``` you should be good

File: master/pkg/model/model.go [hunk 18]
```diff
@@ -6,36 +6,21 @@ import (
...
 	Name            string    `db:"name" json:"name"`
 	Description     string    `db:"description" json:"description"`
-	Metadata        JSONObj   `db:"metadata" json:"metadata"`
 	CreationTime    time.Time `db:"creation_time" json:"creation_time"`
 	LastUpdatedTime time.Time `db:"last_updated_time" json:"last_updated_time"`
+	Metadata        JSONObj   `db:"metadata" json:"metadata"`
+	ID              int       `db:"id" json:"id"`
 	Labels          []string  `db:"labels" json:"labels"`
-	Username        string    `db:"username" json:"username"`
+	UserID          int       `db:"user_id" json:"user_id"`
```
Comment (NicholasBlaskey) [inline master/pkg/model/model.go:18]:
```suggestion
	UserID          UserID       `db:"user_id" json:"user_id"`
```
we have a userID type of ```model.UserID```

File: master/internal/api_workspace.go [hunk 19]
```diff
@@ -551,11 +561,20 @@ func (a *apiServer) DeleteWorkspace(
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to check workspace for models: %w", err)
+	}
+	if modelsExist {
+		return nil, status.Errorf(codes.FailedPrecondition, "workspace (%d) contains models; move or delete models first",
+			req.Id)
+	}
+
 	holder := &workspacev1.Workspace{}
+	// TODO(kristine): update workspace state in transaction with template delete
```
[the remark this answers is not in the document]
Comment (kkunapuli) [reply master/internal/api_workspace.go:574]: Ticket made - https://hpe-aiatscale.atlassian.net/browse/DET-10138

[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply master/internal/api_workspace.go:574]: Could you add the ticket number to this comment?
